File: packages/core/test/engine.test.ts
import { describe, expect, it } from 'vitest';
import {
  compileScript,
  EditEvent,
  humanize,
  parseDirectives,
  Player,
} from '../src/index';

/** Minimal in-memory "editor": inserts at end, backspaces from end. */
function makeBuffer() {
  let text = '';
  const apply = (event: EditEvent) => {
    if (event.kind === 'insert') {
      text += event.text;
    } else {
      text = text.slice(0, text.length - event.count);
    }
  };
  return { apply, get text() { return text; } };
}

const instant = () => Promise.resolve();

describe('parseDirectives', () => {
  it('strips directives and keeps the rest byte-exact', () => {
    const source = 'line1\n//~ pause 800\nline2\n#~ checkpoint intro\nline3\n';
    const segments = parseDirectives(source);
    expect(segments).toEqual([
      { kind: 'text', text: 'line1\n' },
      { kind: 'pause', ms: 800 },
      { kind: 'text', text: 'line2\n' },
      { kind: 'checkpoint', name: 'intro' },
      { kind: 'text', text: 'line3\n' },
    ]);
  });

  it('does not treat regular comments as directives', () => {
    const source = '// a normal comment\n# ~ not glued, not a directive\n';
    expect(parseDirectives(source)).toEqual([{ kind: 'text', text: source }]);
  });
});

describe('humanize', () => {
  it('reproduces the input exactly when typos are disabled', () => {
    const text = 'function add(a, b) {\n  return a + b;\n}\n';
    const steps = humanize(text, { typoRate: 0, seed: 42 });
    const buffer = makeBuffer();
    steps.forEach((s) => buffer.apply(s.event));
    expect(buffer.text).toBe(text);
    expect(steps).toHaveLength([...text].length);
  });

  it('still converges to the input when every letter gets a typo', () => {
    const text = 'const ghost = "typeghost";\n';
    const steps = humanize(text, { typoRate: 1, seed: 7 });
    const buffer = makeBuffer();
    steps.forEach((s) => buffer.apply(s.event));
    expect(buffer.text).toBe(text);
    expect(steps.length).toBeGreaterThan([...text].length);
  });

  it('is deterministic for a given seed', () => {
    const a = humanize('hello world', { seed: 123 });
    const b = humanize('hello world', { seed: 123 });
    expect(a).toEqual(b);
  });
});

describe('compileScript', () => {
  it('produces a clean source without directives', () => {
    const source = '//~ checkpoint\nconsole.log(1);\n//~ pause 500\nconsole.log(2);\n';
    const script = compileScript(source, { typoRate: 0, seed: 1 });
    expect(script.cleanSource).toBe('console.log(1);\nconsole.log(2);\n');
    expect(script.checkpoints).toEqual([{ index: 0, name: undefined }]);
  });
});

describe('Player', () => {
  it('applies the whole script in auto mode', async () => {
    const source = 'const x = 1;\n';
    const script = compileScript(source, { typoRate: 0, seed: 5 });
    const buffer = makeBuffer();
    const player = new Player(script, buffer.apply, { timeScale: 0, sleep: instant });
    await player.play();
    expect(buffer.text).toBe(source);
    expect(player.state).toBe('done');
  });

  it('waits at checkpoints until resumed', async () => {
    const source = 'a\n//~ checkpoint half\nb\n';
    const script = compileScript(source, { typoRate: 0, seed: 5 });
    const buffer = makeBuffer();
    const player = new Player(script, buffer.apply, { timeScale: 0, sleep: instant });

    const playing = player.play();
    await new Promise((r) => setTimeout(r, 20));
    expect(player.state).toBe('waiting-checkpoint');
    expect(buffer.text).toBe('a\n');

    player.resume();
    await playing;
    expect(buffer.text).toBe('a\nb\n');
  });

  it('can be paused and resumed', async () => {
    const source = 'abcdef';
    const script = compileScript(source, { typoRate: 0, seed: 5 });
    const buffer = makeBuffer();
    const player = new Player(script, buffer.apply, { timeScale: 0, sleep: instant });

    player.pause(); // pausing before play() is a no-op
    const playing = player.play();
    player.pause();
    await new Promise((r) => setTimeout(r, 20));
    const frozen = buffer.text;
    await new Promise((r) => setTimeout(r, 20));
    expect(buffer.text).toBe(frozen);

    player.resume();
    await playing;
    expect(buffer.text).toBe(source);
  });

  it('advances manually in hacker mode, finishing typo bursts atomically', async () => {
    const text = 'abc';
    const steps = humanize(text, { typoRate: 1, seed: 9 });
    const buffer = makeBuffer();
    const player = new Player(steps, buffer.apply);

    let alive = true;
    while (alive) {
      alive = await player.advance(1);
      // A typo burst must never leave a dangling wrong character.
      expect(text.startsWith(buffer.text)).toBe(true);
    }
    expect(buffer.text).toBe(text);
    expect(player.state).toBe('done');
  });

  it('catches up when the host throttles timers (hidden-tab survival)', async () => {
    // Simulate Chrome clamping every setTimeout to 1s: the clock jumps 1000ms
    // per sleep, no matter what was requested.
    let clock = 0;
    let sleepCalls = 0;
    const sleep = () => {
      sleepCalls++;
      clock += 1000;
      return Promise.resolve();
    };
    const steps = Array.from({ length: 40 }, (): { event: EditEvent; delayMs: number } => ({
      event: { kind: 'insert', text: 'x' },
      delayMs: 50,
    }));
    const buffer = makeBuffer();
    const player = new Player(steps, buffer.apply, { sleep, now: () => clock });

    await player.play();
    // 40 × 50ms = 2s of virtual typing; with 1s throttled ticks the player
    // must burst-apply overdue steps instead of sleeping 40 times.
    expect(buffer.text).toBe('x'.repeat(40));
    expect(sleepCalls).toBeLessThanOrEqual(4);
  });

  it('serializes overlapping advance calls (rapid keystrokes, async apply)', async () => {
    const text = 'abcdefghijkl';
    const steps = humanize(text, { typoRate: 0, seed: 3 });
    let buffer = '';
    const slowApply = async (event: EditEvent) => {
      await new Promise((r) => setTimeout(r, 1));
      if (event.kind === 'insert') buffer += event.text;
      else buffer = buffer.slice(0, buffer.length - event.count);
    };
    const player = new Player(steps, slowApply);

    // Fire keystrokes without waiting for each other, like a real mash.
    await Promise.all([player.advance(3), player.advance(3), player.advance(3), player.advance(3)]);
    expect(buffer).toBe(text);
  });

  it('reports being exhausted', async () => {
    const steps = humanize('x', { typoRate: 0, seed: 1 });
    const player = new Player(steps, () => undefined);
    expect(await player.advance(10)).toBe(false);
    expect(await player.advance(10)).toBe(false);
  });
});

File: packages/jetbrains-plugin/src/test/kotlin/com/typeghost/engine/EngineTest.kt
package com.typeghost.engine

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/** Minimal in-memory "editor": inserts at end, backspaces from end. */
private class Buffer {
    var text = ""
        private set

    fun apply(event: EditEvent) {
        when (event) {
            is EditEvent.Insert -> text += event.text
            is EditEvent.Backspace -> text = text.dropLast(event.count)
        }
    }
}

/** Fires every task synchronously, advancing a virtual clock by the requested delay. */
private class ImmediateScheduler : Scheduler {
    var clock = 0L
    var scheduleCount = 0

    override fun schedule(delayMs: Long, task: () -> Unit): Cancellable {
        scheduleCount++
        clock += delayMs
        task()
        return Cancellable { }
    }
}

/** Queues tasks so tests can interleave pause/resume between steps. */
private class QueueScheduler : Scheduler {
    var clock = 0L
    private val tasks = ArrayDeque<Pair<Long, () -> Unit>>()

    override fun schedule(delayMs: Long, task: () -> Unit): Cancellable {
        val entry = delayMs to task
        tasks += entry
        return Cancellable { tasks.remove(entry) }
    }

    fun runNext(): Boolean {
        val (delay, task) = tasks.removeFirstOrNull() ?: return false
        clock += delay
        task()
        return true
    }

    fun drain() {
        while (runNext()) Unit
    }
}

class ScriptParserTest {
    @Test
    fun `strips directives and keeps the rest byte-exact`() {
        val source = "line1\n//~ pause 800\nline2\n#~ checkpoint intro\nline3\n"
        val segments = ScriptParser.parseDirectives(source)
        assertEquals(
            listOf(
                Segment.Text("line1\n"),
                Segment.Pause(800),
                Segment.Text("line2\n"),
                Segment.CheckpointMark("intro"),
                Segment.Text("line3\n"),
            ),
            segments,
        )
    }

    @Test
    fun `does not treat regular comments as directives`() {
        val source = "// a normal comment\n# ~ not glued, not a directive\n"
        assertEquals(listOf<Segment>(Segment.Text(source)), ScriptParser.parseDirectives(source))
    }

    @Test
    fun `compile produces a clean source and checkpoint indices`() {
        val source = "//~ checkpoint\nconsole.log(1);\n//~ pause 500\nconsole.log(2);\n"
        val script = ScriptParser.compile(source, HumanizeOptions(typoRate = 0.0, seed = 1))
        assertEquals("console.log(1);\nconsole.log(2);\n", script.cleanSource)
        assertEquals(listOf(Checkpoint(0, null)), script.checkpoints)
    }
}

class HumanizerTest {
    @Test
    fun `reproduces the input exactly when typos are disabled`() {
        val text = "function add(a, b) {\n  return a + b;\n}\n"
        val steps = Humanizer.humanize(text, HumanizeOptions(typoRate = 0.0, seed = 42))
        val buffer = Buffer()
        steps.forEach { buffer.apply(it.event) }
        assertEquals(text, buffer.text)
        assertEquals(text.codePointCount(0, text.length), steps.size)
    }

    @Test
    fun `still converges to the input when every letter gets a typo`() {
        val text = "const ghost = \"typeghost\";\n"
        val steps = Humanizer.humanize(text, HumanizeOptions(typoRate = 1.0, seed = 7))
        val buffer = Buffer()
        steps.forEach { buffer.apply(it.event) }
        assertEquals(text, buffer.text)
        assertTrue(steps.size > text.length)
    }

    @Test
    fun `is deterministic for a given seed`() {
        val a = Humanizer.humanize("hello world", HumanizeOptions(seed = 123))
        val b = Humanizer.humanize("hello world", HumanizeOptions(seed = 123))
        assertEquals(a, b)
    }
}

class PlayerTest {
    private fun compiled(source: String) =
        ScriptParser.compile(source, HumanizeOptions(typoRate = 0.0, seed = 5))

    @Test
    fun `applies the whole script in auto mode`() {
        val source = "const x = 1;\n"
        val buffer = Buffer()
        val scheduler = ImmediateScheduler()
        val player = Player(compiled(source), buffer::apply, scheduler, now = { scheduler.clock })
        player.play()
        assertEquals(source, buffer.text)
        assertEquals(PlayerState.DONE, player.state)
    }

    @Test
    fun `waits at checkpoints until resumed`() {
        val source = "a\n//~ checkpoint half\nb\n"
        val buffer = Buffer()
        val scheduler = ImmediateScheduler()
        val player = Player(compiled(source), buffer::apply, scheduler, now = { scheduler.clock })

        player.play()
        assertEquals(PlayerState.WAITING_CHECKPOINT, player.state)
        assertEquals("a\n", buffer.text)

        player.resume()
        assertEquals("a\nb\n", buffer.text)
        assertEquals(PlayerState.DONE, player.state)
    }

    @Test
    fun `can be paused and resumed`() {
        val source = "abcdef"
        val buffer = Buffer()
        val scheduler = QueueScheduler()
        val player = Player(compiled(source), buffer::apply, scheduler, now = { scheduler.clock })

        player.pause() // pausing before play() is a no-op
        player.play()
        scheduler.runNext()
        scheduler.runNext()
        player.pause()
        val frozen = buffer.text
        assertFalse(scheduler.runNext(), "pause must cancel the pending timer")
        assertEquals(frozen, buffer.text)

        player.resume()
        scheduler.drain()
        assertEquals(source, buffer.text)
        assertEquals(PlayerState.DONE, player.state)
    }

    @Test
    fun `catches up when the host throttles timers`() {
        // Simulate a clamped timer: the clock jumps 1000ms per schedule,
        // no matter what was requested.
        val steps = List(40) { Step(EditEvent.Insert("x"), 50) }
        val script = CompiledScript(steps, emptyList(), "x".repeat(40))
        val buffer = Buffer()
        var clock = 0L
        var scheduleCount = 0
        val throttled = Scheduler { _, task ->
            scheduleCount++
            clock += 1000
            task()
            Cancellable { }
        }
        val player = Player(script, buffer::apply, throttled, now = { clock })
        player.play()
        // 40 × 50ms = 2s of virtual typing; with 1s throttled ticks the player
        // must burst-apply overdue steps instead of scheduling 40 times.
        assertEquals("x".repeat(40), buffer.text)
        assertTrue(scheduleCount <= 4, "expected few schedules, got $scheduleCount")
    }

    @Test
    fun `advances manually, finishing typo bursts atomically`() {
        val text = "abc"
        val steps = Humanizer.humanize(text, HumanizeOptions(typoRate = 1.0, seed = 9))
        val script = CompiledScript(steps, emptyList(), text)
        val buffer = Buffer()
        val player = Player(script, buffer::apply, ImmediateScheduler())

        var alive = true
        while (alive) {
            alive = player.advance(1)
            // A typo burst must never leave a dangling wrong character.
            assertTrue(text.startsWith(buffer.text), "unexpected buffer: '${buffer.text}'")
        }
        assertEquals(text, buffer.text)
        assertEquals(PlayerState.DONE, player.state)
    }

    @Test
    fun `reports being exhausted`() {
        val steps = Humanizer.humanize("x", HumanizeOptions(typoRate = 0.0, seed = 1))
        val player = Player(CompiledScript(steps, emptyList(), "x"), {}, ImmediateScheduler())
        assertFalse(player.advance(10))
        assertFalse(player.advance(10))
    }
}
